.buildkite/scripts/changesets.sh

```diff
@@ -1,77 +1,62 @@
   fi
 }
 
-are_changed_only_paths() {
-  local patterns=("${@}")
-  local changed_files=($(git diff --name-only HEAD@{1} HEAD))
-  local matched_files=()
-  for pattern in "${patterns[@]}"; do
-    local matched=($(grep -E "${pattern}" <<< "${changed_files[@]}"))
-    if [ "${#matched[@]}" -gt 0 ]; then
-      matched_files+=("${matched[@]}")
+defineExclusions() {
+  exclude="^$beatPath\/module\/(.*(?<!\.asciidoc|\.png))$"
+}
+
+defineFromCommit() {
+  local previousCommit
+  local changeTarget=${BUILDKITE_PULL_REQUEST_BASE_BRANCH:-$BUILDKITE_BRANCH}
+
+  previousCommit=$(git rev-parse HEAD^)
+
+  from=${changeTarget:+"origin/$changeTarget"}
+  from=${from:-$previousCommit}
+  from=${from:-$BUILDKITE_COMMIT}
+}
+
+getMatchingModules() {
+  local changedPaths
+  mapfile -t changedPaths < <(git diff --name-only "$from"..."$BUILDKITE_COMMIT" | grep -P "$exclude" | grep -oP "$pattern")
+  mapfile -t modulesMatched < <(printf "%s\n" "${changedPaths[@]}" | grep -o 'module/.*' | awk -F '/' '{print $2}' | sort -u)
+}
+
+addToModuleEnvVar() {
+  local module
+  for module in "${modulesMatched[@]}"; do
+    if [[ -z ${modules+x} ]]; then
+      modules="$module"
+    else
+      modules+=",$module"
     fi
   done
-  if [ "${#matched_files[@]}" -eq "${#changed_files[@]}" ] || [ "${#changed_files[@]}" -eq 0 ]; then
-    return 0
-  fi
-  return 1
 }
 
 defineModuleFromTheChangeSet() {
-  # This function sets a `MODULE` env var, required by IT tests, containing a comma separated list of modules for a given beats project (specified via the first argument).
-  # The list is built depending on directories that have changed under `modules/` excluding anything else such as asciidoc and png files.
-  # `MODULE` will empty if no changes apply.
-  local project_path=$1
-  local project_path_transformed=$(echo "$project_path" | sed 's/\//\\\//g')
-  local project_path_exclussion="((?!^${project_path_transformed}\\/).)*\$"
-  local exclude=("^(${project_path_exclussion}|((?!\\/module\\/).)*\$|.*\\.asciidoc|.*\\.png)")
+  beatPath=$1
 
-  if [[ "$project_path" == *"x-pack/"* ]]; then
-    local pattern=("$XPACK_MODULE_PATTERN")
-  else
-    local pattern=("$OSS_MODULE_PATTERN")
-  fi
-  local changed_modules=""
-  local module_dirs=$(find "$project_path/module" -mindepth 1 -maxdepth 1 -type d)
-  for module_dir in $module_dirs; do
-    if are_paths_changed $module_dir && ! are_changed_only_paths "${exclude[@]}"; then
-      if [[ -z "$changed_modules" ]]; then
-        changed_modules=$(basename "$module_dir")
-      else
-        changed_modules+=",$(basename "$module_dir")"
-      fi
-    fi
-  done
+  definePattern
+  defineExclusions
+  defineFromCommit
+  getMatchingModules
 
-  # export MODULE="" leads to an infinite loop https://github.com/elastic/ingest-dev/issues/2993
-  if [[ ! -z $changed_modules ]]; then 
-    export MODULE="${changed_modules}"
-    echo "~~~ Set env var MODULE to [$MODULE]"
-    echo "~~~ Resuming commands"
```

Comment:
I think it's useful to keep debug messages like the above to help the author understand what's being run.

e.g. something along the lines of 

```
echo "~~~ Detected file changes for some modules. Setting env var MODULE to [$MODULE]"
echo "~~~ Resuming commands"
```

or something along those line